### PR TITLE
DIGREPO-777: generate thumbnails during ingest

### PR DESCRIPTION
**Pull request by dunn**
Don't merge; still testing.

### CODE REVIEW

**Comment by dunn** [comment]
Needs to be turned on for maps too.

**Comment by dunn** [comment]
For some reason the square thumbnails aren't being generated correctly (e.g., http://localhost:3000/image-service/c247ds92f%2Ffiles%2F1049a01c-997f-4332-82df-7f9f4b419d44/square/100,/0/default.jpg); so RIIIF is making them again on request.

File: lib/importer/factory/object_factory.rb [hunk 1]
```diff
@@ -136,6 +81,85 @@ def create
...
+    #
+    # @param [Hash] object
+    # @param [String, Hash] files Either the path to the file or a
+    #     hash with file metadata
+    def attach_files(_object, _files)
+      raise NotImplementedError,
+            "#attach_files is not defined for #{self.class}"
+    end
+
+    def render_thumbnails(object)
```
Comment (dunn) [inline lib/importer/factory/object_factory.rb]: @supritasrinivas I organized ObjectFactory so the diff is a little confusing; here's the new method.

**Comment by dunn** [comment]
Here's the problem.  We've created a square (cropped) and a rectangular (uncropped) thumbnail for each image.  Now we are asking RIIIF to give us the square one.  It needs to figure out the filename of that thumbnail, so it can load it into memory.

The filename is a hash of the File ID along with the options RIIIF used to generate it: size, rotation, region (crop), etc.: https://github.com/curationexperts/riiif/blob/0e891ae000b79e4b44fae659fbf815bd2f206efc/app/models/riiif/image.rb#L73-L78

The value for `region` that we used to generate the thumbnail was 'square'.  But unfortunately ImageMagick doesn't know what 'square' means.  So RIIIF converts it to something like `3600x3600+0+0`; the first part means to size it at 3600x3600 (the full length or width, whichever is smaller), and the second part means that ImageMagick shouldn't trim the edges at all.

So what _that_ means is that in order to compute the filename of the cached thumbnail, RIIIF needs to know the dimensions of the original image.  We have that information indexed in Solr, but RIIIF doesn't assume anything, so it re-reads the original image and get the dimensions directly.

You see where this is going.  To re-read the original image we need to download it again from Fedora.  Which is why pulling square thumbnails out of the cache ends up downloading them again.

**Comment by supritasrinivas** [comment]
Nice Sleuthing on the duplicated image downloads mystery!

**Comment by dunn** [comment]
https://github.com/curationexperts/riiif/pull/85

**Comment by supritasrinivas** [comment]
@dunn Is this ready to be merged?

**Comment by dunn** [comment]
Yes, sorry for the delay.  I only just now found a place with reliable wifi.

**Comment by dunn** [comment]
Oh ugh, now we need ImageMagick on the Jenkins workers: http://jenkins.library.ucsb.edu:8080/job/ADRL_master/341/console

**Pull request merged**